pkg/kubelet/status/status_manager.go

```diff
@@ -711,23 +697,29 @@ func (m *manager) RemoveOrphanedStatuses(podUIDs map[types.UID]bool) {
 }
 
 // syncBatch syncs pods statuses with the apiserver.
-func (m *manager) syncBatch() {
+func (m *manager) syncBatch(all bool) {
 	var updatedStatuses []podStatusSyncRequest
 	podToMirror, mirrorToPod := m.podManager.GetUIDTranslations()
 	func() { // Critical section
 		m.podStatusesLock.RLock()
 		defer m.podStatusesLock.RUnlock()
 
 		// Clean up orphaned versions.
-		for uid := range m.apiStatusVersions {
-			_, hasPod := m.podStatuses[types.UID(uid)]
-			_, hasMirror := mirrorToPod[uid]
-			if !hasPod && !hasMirror {
-				delete(m.apiStatusVersions, uid)
+		if all {
+			for uid := range m.apiStatusVersions {
+				_, hasPod := m.podStatuses[types.UID(uid)]
+				_, hasMirror := mirrorToPod[uid]
+				if !hasPod && !hasMirror {
+					delete(m.apiStatusVersions, uid)
+				}
 			}
 		}
 
 		for uid, status := range m.podStatuses {
+			// skip non updated statuses
+			if _, ok := m.podStatusesNew[uid]; !all && !ok {
```

Comment:
for my understanding, if `all` is false, why do want to update statuses even if the status is not in `podStatusesNew`? Doesn't this mean that the status is unchanged?

Reply:
Maybe the naming is confusing (i don't like this sort of if block but the alternatives were worse)

```
func syncBatch(allStatusVsJustChangedStatus) {
   ...
   for uid, status := range allPodStatuses {
      // if we're not syncing all pods, check to see whether this pod has changed before deciding to sync it
      ...
      // queue this pod to sync
   }
   ...
}
```

`podStatusNew` could be called `uidsOfPodsWithUpdatedStatuses`

Reply:
That makes sense, I think I have broader question here: What is the purpose of having this logic of `podStatusesNew` and only using it for the non batch case? 

As I understand, for every status update we check `m.podStatuses` (desired status) against the cached api status (`apiStatusVersions` from pod manager) and only update the status if there is a diff (which is checked via `needsReconcile` and `needsUpdate`). 

If that's the case, why do we need logic of `podStatusesNew` instead of just relying on if there is a diff?

Reply:
We could.  It's potentially more churn, but I think it makes sense.  If you'd like that I'll make the change.

Reply:
Hrm, this makes me realize there is more bandaid here (this is what I bogged down in changing last time).  NeedsUpdate and NeedsReconcile should be removed and refactored, but i didn't want to do it here because of more risk.  I think it's safe for apiStatusVersions to be the check though...

Reply:
Makes sense... I don't have explicit preference, it just seems that if we removed  `podStatusesNew` and the relied on the diff it would simplify the logic here a bit and the code path for both batch and non-batch updates would be largely the same.

Reply:
Super super subtle behavior change if we use apiStatusVersions - if a failure happens, we retry that failure the next time any pod has a status update, instead of at the next batch.  If the patch cannot be applied for some reason, the kubelet would retry that on every status change of any pod, instead of every syncPeriod (10s). 

Somewhat concerning, but not very concerning, because any pod like that is still broken, and the real way to fix it is to track that deliberately.  Also, I don't think we've ever had that problem (except maybe on an unsupported node skew) and you're broken in those cases anyway.  I think it's the right change.  Pushed an update and added more comments to these methods.

Reply:
Makes sense, and thanks for the callout of the subtle difference. I agree, it sounds like despite the change it shouldn't have any negative effect.

Reply:
I think the north star here to get away from is a single threaded status loop with deep queueing.  We want to dispatch faster, simpler, individual calls.